From 54e1392c26b3e54cfddc7d52ea81f0ea550a1270 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sat, 6 Mar 2010 16:35:52 +0000
Subject: [PATCH] CONFIGURATION-409: HierarchicalINIConfiguration.save() now
 works correctly with sections containing a delimiter character in their name.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@919796 13f79535-47bb-0310-9956-ffa450edef68
---
 .../HierarchicalINIConfiguration.java           |  7 ++++++-
 .../TestHierarchicalINIConfiguration.java       | 17 +++++++++++++++++
 xdocs/changes.xml                               |  4 ++++
 3 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/src/java/org/apache/commons/configuration/HierarchicalINIConfiguration.java b/src/java/org/apache/commons/configuration/HierarchicalINIConfiguration.java
index 4ac5528da7..905f52e116 100644
--- a/src/java/org/apache/commons/configuration/HierarchicalINIConfiguration.java
+++ b/src/java/org/apache/commons/configuration/HierarchicalINIConfiguration.java
@@ -269,15 +269,20 @@ public void save(Writer writer) throws ConfigurationException
         while (it.hasNext())
         {
             String section = (String) it.next();
+            Configuration subset;
             if (section != null)
             {
                 out.print("[");
                 out.print(section);
                 out.print("]");
                 out.println();
+                subset = createSubnodeConfiguration(getSectionNode(section));
+            }
+            else
+            {
+                subset = getSection(null);
             }
 
-            Configuration subset = getSection(section);
             Iterator keys = subset.getKeys();
             while (keys.hasNext())
             {
diff --git a/src/test/org/apache/commons/configuration/TestHierarchicalINIConfiguration.java b/src/test/org/apache/commons/configuration/TestHierarchicalINIConfiguration.java
index 7041fd5187..edaf639a93 100644
--- a/src/test/org/apache/commons/configuration/TestHierarchicalINIConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestHierarchicalINIConfiguration.java
@@ -630,6 +630,23 @@ public void testLineContinuationAtEnd() throws ConfigurationException
                 .getString("section5.continueNoLine"));
     }
 
+    /**
+     * Tests whether a configuration can be saved that contains section keys
+     * with delimiter characters. This test is related to CONFIGURATION-409.
+     */
+    public void testSaveKeysWithDelimiters() throws ConfigurationException
+    {
+        HierarchicalINIConfiguration conf = new HierarchicalINIConfiguration();
+        final String section = "Section..with..dots";
+        conf.addProperty(section + ".test1", "test1");
+        conf.addProperty(section + ".test2", "test2");
+        conf.save(TEST_FILE);
+        conf = new HierarchicalINIConfiguration();
+        conf.load(TEST_FILE);
+        assertEquals("Wrong value (1)", "test1", conf.getString(section + ".test1"));
+        assertEquals("Wrong value (2)", "test2", conf.getString(section + ".test2"));
+    }
+
     /**
      * A thread class for testing concurrent access to the global section.
      */
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 0815020774..8b93a8269a 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -23,6 +23,10 @@
 
   <body>
     <release version="1.7" date="in SVN" description="">
+      <action dev="oheger" type="fix" issue="CONFIGURATION-409">
+        HierarchicalINIConfiguration now correctly saves sections whose name
+        contains delimiter characters.
+      </action>
       <action dev="oheger" type="fix" issue="CONFIGURATION-407">
         Fixed a potential IllegalStateException in HierarchicalINIConfiguration
         that can be thrown when the global section is requested concurrently.
